nvflare/apis/impl/client_controller.py

```diff
@@ -0,0 +1,223 @@
+            # if task_cb_error:
+            #     return self._make_error_reply(ReturnCode.ERROR, targets)
+
+        request.set_header(ReservedKey.TASK_NAME, task.name)
+        replies = engine.send_aux_request(
+            targets=targets, topic=ReservedTopic.DO_TASK, request=request, timeout=task.timeout, fl_ctx=fl_ctx
+        )
+
+        self.log_debug(fl_ctx, "firing event EventType.AFTER_TASK_EXECUTION")
+        self.fire_event(EventType.AFTER_TASK_EXECUTION, fl_ctx)
+
+        self.log_debug(fl_ctx, "firing event EventType.BEFORE_TASK_RESULT_FILTER")
+        self.fire_event(EventType.BEFORE_TASK_RESULT_FILTER, fl_ctx)
+
+        task_filter_list = self.task_result_filters.get(task.name)
+        for target, reply in replies.items():
+            # get the client task for the target
+            for client_task in task.client_tasks:
+                if client_task.client.name == target:
+                    # apply result filters
+                    filter_error, result = self.apply_result_filters(task_filter_list, reply, fl_ctx)
+                    if filter_error:
+                        error_reply = make_reply(ReturnCode.TASK_RESULT_FILTER_ERROR)
+                        client_task.result = error_reply
+                        break
+
+                    # assign replies to client task, prepare for the result_received_cb
+                    client_task.result = reply
+
+                    client: Client = self._get_client(target, engine)
+                    task_cb_error = self._call_task_cb(task.result_received_cb, client, task, fl_ctx)
+                    if task_cb_error:
+                        client_task.result = make_reply(ReturnCode.ERROR)
+                        break
+
+                    break
+
+        # apply task_done_cb
+        if task.task_done_cb is not None:
+            try:
+                task.task_done_cb(task=task, fl_ctx=fl_ctx)
+            except Exception as e:
+                self.log_exception(
+                    fl_ctx, f"processing error in task_done_cb error on task {task.name}: {secure_format_exception(e)}"
+                ),
+                task.completion_status = TaskCompletionStatus.ERROR
+                task.exception = e
+                return self._make_error_reply(ReturnCode.ERROR, targets)
+
+        replies = {}
+        for client_task in task.client_tasks:
+            replies[client_task.client.name] = client_task.result
+        return replies
+
+    def _make_error_reply(self, error_type, targets):
+        error_reply = make_reply(error_type)
+        replies = {}
+        for target in targets:
+            replies[target] = error_reply
+        return replies
+
+    def _get_client(self, client, engine) -> Client:
+        if isinstance(client, Client):
+            return client
+
+        if client == SiteType.SERVER:
+            return Client(SiteType.SERVER, None)
```

Comment:
This is weird passing "SiteType.SERVER" to class "Client".

Reply:
The Client is more accurate to be called Site, so the Server is treated one of them. Since Client has been used in a lot of places, that will be a much bigger change scope. We will leave that to future change.